### PR TITLE
Take the full funding transaction from the user on generation

### PR DESCRIPTION
**Pull request by TheBlueMatt**
Instead of relying on the user to ensure the funding transaction is
correct (and panicing when it is confirmed), we should check it is
correct when it is generated. By taking the full funding transaciton
from the user on generation, we can also handle broadcasting for
them instead of doing so via an event.

### CODE REVIEW

**Comment by ariard** [comment]
Concept ACK, will be needed for dual-funding/splicing support anyway.

File: lightning/src/ln/channelmanager.rs [hunk 1]
```diff
@@ -1528,29 +1529,48 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
...
 				},
-				None => return
+				None => { return Err(APIError::ChannelUnavailable { err: "No such channel".to_owned() }) },
 			};
 			match handle_error!(self, res, chan.get_counterparty_node_id()) {
 				Ok(funding_msg) => {
 					(chan, funding_msg)
 				},
-				Err(_) => { return; }
+				Err(_) => { return Err(APIError::ChannelUnavailable { err: "Error deriving keys - either our RNG or our counterparty's RNG is broken".to_owned() }) },
```
Comment (valentinewallace) [inline lightning/src/ln/channelmanager.rs:1573]: Hmm, couldn't it have been another error? Seems that we can reach this point only if `sign_counterparty_commitment` fails

**Review by valentinewallace — COMMENTED** [review]
This looks super reasonable to me! I love API improvements like this. Let's get rid of `PendingHTLCsForwardable` next :zany_face:

File: lightning/src/ln/channelmanager.rs [hunk 2]
```diff
@@ -1528,29 +1529,48 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
...
 	///
+	/// Returns an [`APIError::APIMisuseError`] if the funding_transaction spent non-SegWit outputs
+	/// or the output didn't match the parameters in [`Event::FundingGenerationReady`].
+	///
 	/// Panics if a funding transaction has already been provided for this channel.
 	///
 	/// May panic if the funding_txo is duplicative with some other channel (note that this should
 	/// be trivially prevented by using unique funding transaction keys per-channel).
-	pub fn funding_transaction_generated(&self, temporary_channel_id: &[u8; 32], funding_txo: OutPoint) {
+	pub fn funding_transaction_generated(&self, temporary_channel_id: &[u8; 32], funding_transaction: Transaction, output_index: u16) -> Result<(), APIError> {
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:1539]: I guess in theory we could walk the transaction to find one with a matching script, do you feel strongly that we should?

File: lightning/src/ln/channelmanager.rs (hunk 1 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:1577]: Hmm, I was looking at the `build_remote_transaction_keys()` failures, though you're right, `sign_counterparty_commitment` is a user function, so it could be for any reason (including hardware wallet disconnected). If its our own `InMemorySigner`, I think that's still the only case.

File: lightning/src/ln/channelmanager.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (valentinewallace) [reply lightning/src/ln/channelmanager.rs:1544]: I think that'd be a big win since it's a user-facing function :)

File: lightning/src/ln/channel.rs [hunk 3]
```diff
@@ -2771,20 +2773,22 @@ impl<Signer: Sign> Channel<Signer> {
...
 
-		// Because we will never generate a FundingBroadcastSafe event when we're in
-		// MonitorUpdateFailed, if we assume the user only broadcast the funding transaction when
-		// they received the FundingBroadcastSafe event, we can only ever hit
+		// Because we will never broadcast the funding transaction when we're in
+		// MonitorUpdateFailed, if we assume the user never directly broadcasts the funding
+		// transaction and waits for us to do it, we can only ever hit
 		// monitor_pending_funding_locked when we're an inbound channel which failed to persist the
 		// monitor on funding_created, and we even got the funding transaction confirmed before the
 		// monitor was persisted.
```
Comment (valentinewallace) [inline lightning/src/ln/channel.rs:2789]: Hm, this comment is all one sentence, maybe could be broken up a bit. Also, IIUC, maybe rephrase "we even got the funding transaction confirmed" to "the funding transaction was even confirmed before the monitor..." (since presumably the counterparty broadcasted it, not us?).

File: lightning/src/ln/channel.rs [hunk 4]
```diff
@@ -2815,11 +2819,11 @@ impl<Signer: Sign> Channel<Signer> {
 		self.monitor_pending_commitment_signed = false;
 		let order = self.resend_order.clone();
 		log_trace!(logger, "Restored monitor updating resulting in {}{} commitment update and {} RAA, with {} first",
-			if needs_broadcast_safe { "a funding broadcast safe, " } else { "" },
+			if funding_broadcastable.is_some() { "a funding broadcast safe, " } else { "" },
```
Comment (valentinewallace) [inline lightning/src/ln/channel.rs:2822]: nit: s/a funding broadcast safe/a broadcastable funding transaction

**Review by valentinewallace — APPROVED** [review]
Just some nits, the code LGTM :100:

File: lightning/src/ln/channel.rs [hunk 5]
```diff
@@ -2771,20 +2773,21 @@ impl<Signer: Sign> Channel<Signer> {
...
-		// monitor on funding_created, and we even got the funding transaction confirmed before the
-		// monitor was persisted.
+		// We will never broadcast the funding transaction when we're in MonitorUpdateFailed (and
+		// we assume the user never directly broadcasts the funding transaction and waits for us to
+		// do it). Thus, we can only ever hit monitor_pending_funding_locked when we're an inbound
+		// channel which failed to persist the monitor on funding_created, and we got the funding
+		// transaction confirmed before the monitor was persisted.
 		let funding_locked = if self.monitor_pending_funding_locked {
-			assert!(!self.is_outbound(), "Funding transaction broadcast without FundingBroadcastSafe!");
+			assert!(!self.is_outbound(), "Funding transaction broadcast manually - we didn't do it!");
```
Comment (ariard) [inline lightning/src/ln/channel.rs:2790]: Could you precise further to dissociate ambiguity from _manual_ broadcast by counterparty "Funding transaction broadcast manually by user - implementation didn't do it".

File: lightning/src/ln/channelmanager.rs [hunk 6]
```diff
@@ -1528,29 +1530,71 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
...
+	/// Do NOT broadcast the funding transaction yourself. When we have safely received our
+	/// counterparty's signature the funding transaction will automatically be broadcast via the
+	/// [`BroadcasterInterface`] provided when this `ChannelManager` was constructed.
+	pub fn funding_transaction_generated(&self, temporary_channel_id: &[u8; 32], funding_transaction: Transaction) -> Result<(), APIError> {
 		let _persistence_guard = PersistenceNotifierGuard::new(&self.total_consistency_lock, &self.persistence_notifier);
 
+		for inp in funding_transaction.input.iter() {
+			if inp.witness.is_empty() {
+				return Err(APIError::APIMisuseError {
+					err: "Funding transaction must be fully signed and spend Segwit outputs".to_owned()
```
Comment (ariard) [inline lightning/src/ln/channelmanager.rs:1551]: By transaction must be fully signed you meant use only SIGHASH_ALL otherwise you might have in-flight malleability breaking the signatures for the first commitment transaction ? 

Ideally we have the signatures available in the witnesses so in the future we might parse them to verify sighash types flag to protect users against footgunish behaviors.

File: lightning/src/ln/channelmanager.rs [hunk 7]
```diff
@@ -1528,29 +1530,71 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
...
+					let mut output_index = None;
+					let expected_spk = chan.get_funding_redeemscript().to_v0_p2wsh();
+					for (idx, outp) in funding_transaction.output.iter().enumerate() {
+						if outp.script_pubkey == expected_spk && outp.value == chan.get_value_satoshis() {
+							if output_index.is_some() {
+								return Err(APIError::APIMisuseError {
+									err: "Multiple outputs matched the expected script and value".to_owned()
+								});
+							}
+							if idx > u16::max_value() as usize {
```
Comment (ariard) [inline lightning/src/ln/channelmanager.rs:1568]: If you assume minimal-sized output (8 bytes value + 2 bytes scriptPubkey), given that MAX_STANDARD_TX_WEIGHT is about 100k bytes, I think you can just reject a transaction with an idx above 10_000 ?

**Review by ariard — COMMENTED** [review]
Minor comments, otherwise sounds good to me :)

File: lightning/src/ln/channelmanager.rs (hunk 6 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:1551]: Hmm, we do have them here, but we'd need to parse the original script_pubkey, which I don't think we want users to have to include. I can include it in the docs.

File: lightning/src/ln/channelmanager.rs (hunk 7 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:1568]: Sure, but its nice to get an error instead of a panic, even if you did something kinda stupid.

File: lightning/src/ln/channelmanager.rs (hunk 6 above)
Comment (ariard) [reply lightning/src/ln/channelmanager.rs:1548]: If the witness change in the wrong way with a SIGHASH_SINGLE as you point it that's not fine. Should we document "don't construct the transaction such that someone can arbitrarily modify it " ? I would say we suffer from cognitive dissonance there. It's quite obvious to us but in practice users might authorize malleation or even RBF for fee management (see https://lists.linuxfoundation.org/pipermail/lightning-dev/2021-March/002981.html).

**Comment by ariard** [comment]
Code Review b42f594

If you wanna document more [malleation risks](https://github.com/rust-bitcoin/rust-lightning/pull/856#discussion_r610925987), I'm happy.

File: lightning/src/ln/channelmanager.rs (hunk 6 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:1548]: Right, I guess we could re-enforce the last line - if you don't broadcast the transaction yourself, you can't RBF either :).

**Comment by TheBlueMatt** [comment]
Added a note
```
        /// Do NOT broadcast the funding transaction yourself. When we have safely received our
        /// counterparty's signature the funding transaction will automatically be broadcast via the
        /// [`BroadcasterInterface`] provided when this `ChannelManager` was constructed.
+       ///
+       /// Note that this includes RBF or similar transaction replacement strategies - lightning does
+       /// not currently support replacing a funding transaction on an existing channel. Instead,
+       /// create a new channel with a conflicting funding transaction.
```

**Comment by ariard** [comment]
Code Review ACK 26a2f25 Thanks for the comment :)

**Comment by TheBlueMatt** [comment]
Squashed with no changes, will merge after CI.
```
$ git diff-tree -U3 26a2f258 3f2efcdf
$
```
